### PR TITLE
Demonstrate failure of init() functions containing create_entry call

### LINKED ISSUE #599: Seeing DB corruption in CasBufUsedSync::get, trying to do a CapGrant
Intermittently during Javascript testing of a Zome, I see the following failure while trying to do a CapGrant during the Zome's `fn init( _: () )`:

```
★ thread 'holochain-tokio-thread' panicked at 'Holochain detected database corruption.
★
★ Corrupt module: CasBufUsedSync::get
★ Expected hash: [132, 33, 36, 136, 181, 215, 48, 179, 69, 97, 244, 50, 224, 38, 215, 215, 228, 106, 136, 29, 210, 99, 35, 143, 3, 228, 213, 235, 181, 77, 184, 195, 73, 56, 124, 204, 195, 167, 9]
★ Actual hash: [132, 33, 36, 50, 150, 172, 101, 76, 225, 68, 3, 255, 214, 74, 213, 28, 216, 28, 109, 127, 19, 11, 171, 152, 154, 12, 182, 29, 242, 244, 193, 234, 43, 4, 205, 106, 25, 124, 240]
★ Content: CapGrant(ZomeCallCapGrant { tag: "", access: Unrestricted, functions: {(ZomeName("mutualcredit"), FunctionName("authorize")), (ZomeName("mutualcredit"), FunctionName("transfer_authorized"))} }\
)
★
★ We are shutting down as a precaution to prevent further corruption.', /tmp/nix-build-holochain.drv-1/source/crates/holochain_lmdb/src/buffer/cas/buf_sync.rs:107:9
★ stack backtrace:
★
08:21:42 ^[[32minfo^[[39m:
☯☯☯ {{{CONDUCTOR c0}}}
☯    0: rust_begin_unwind
☯    1: std::panicking::begin_panic_fmt
☯    2: holochain_lmdb::buffer::cas::buf_sync::CasBufUsedSync<C,P>::deserialize_and_hash
☯    3: holochain_lmdb::buffer::
08:21:42 ^[[32minfo^[[39m:
☸☸☸ {{{CONDUCTOR c0}}}
☸ cas::buf_sync::CasBufFreshSync<C,P>::get
☸    4: holochain_state::element_buf::ElementBuf<P>::get_entry
☸    5: holochain_state::source_chain::SourceChain::query
☸    6: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
☸    7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
☸    8: std::thread::local::LocalKey<T>::with
☸    9: tokio::runtime::enter::Enter::block_on
☸   10: tokio::runtime::handle::Handle::block_on
☸   11: tokio::runtime::thread_pool::worker::block_in_place
☸   12: holochain::core::ribosome::host_fn::query::query
☸   13: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
☸   14: std::panicking::try
☸   15: <FN as wasmer_runtime_core::typed_func::HostFunction<wasmer_runtime_core::typed_func::
08:21:42 ^[[32minfo^[[39m:
☮☮☮ {{{CONDUCTOR c0}}}
☮ ExplicitVmCtx,A,Rets>>::to_raw::wrap
☮   16: <unknown>
☮ note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
☮
08:21:42 [tryorama] ^[[31merror^[[39m: Test error: {
  type: 'error',
  data: {
    type: 'ribosome_error',
    data: 'Wasm error while working with Ribosome: CallError("RuntimeError: \\"Holochain detected database corruption.\\n\\nCorrupt module: CasBufUsedSync::get\\nExpected hash: [132, 33, 36, 136, 181, 21\
5, 48, 179, 69, 97, 244, 50, 224, 38, 215, 215, 228, 106, 136, 29, 210, 99, 35, 143, 3, 228, 213, 235, 181, 77, 184, 195, 73, 56, 124, 204, 195, 167, 9]\\nActual hash: [132, 33, 36, 50, 150, 172, 101, 76\
, 225, 68, 3, 255, 214, 74, 213, 28, 216, 28, 109, 127, 19, 11, 171, 152, 154, 12, 182, 29, 242, 244, 193, 234, 43, 4, 205, 106, 25, 124, 240]\\nContent: CapGrant(ZomeCallCapGrant { tag: \\"\\", access: \
Unrestricted, functions: {(ZomeName(\\"mutualcredit\\"), FunctionName(\\"authorize\\")), (ZomeName(\\"mutualcredit\\"), FunctionName(\\"transfer_authorized\\"))} })\\n\\nWe are shutting down as a precaut\
ion to prevent further corruption.\\"")'
  }
}
```

### PR DESCRIPTION
**Pull request by pjkundert**
When fixed, this should resolve #599.

This pull request demonstrates that adding an hdk3::prelude::create_entry to a Zome init function will fail, when there are multiple Zomes with init functions that commit to the source-chain.  For example, if they do CapGrants opening up access to Zome APIs in their init (which causes commits to the source-chain).

o When multiple Zomes have init() calls, and some of them perform
  CapGrants and others perform create_entry, failures occur.
o Also added some tests to confirm that CapGrants in init are working,
  when multiple Zomes have init() functions.
o Commenting out the hdk3::prelude::create_entry call around line 112 in
  crates/test_utils/wasm/wasm_workspace/create_entry/src/lib.rs
  causes the tests to pass, otherwise it will fail with various
  errors or panics.

To see these tests fail, check out the 2021-01-26-init-commit-fail branch, and run:

```
nix-shell
cargo build --features 'build_wasms' --manifest-path=crates/holochain/Cargo.toml
RUST_LOG=[debug]=debug TRYORAMA_LOG_LEVEL=info RUST_BACKTRACE=1 cargo test wasm
```

You will see some of the wasm tests fail/panic.

Then, go into crates/test_utils/wasm/wasm_workspace/create_entry/src/lib.rs and comment out the create_entry around line 112; run the same tests, and they will pass.

### CODE REVIEW

**Comment by pjkundert** [comment]
OK, I've discovered one interesting thing.  Using HashSet/Map in committed entries is a no-no -- it doesn't De/Serialize deterministically, so the same entry committed, deserialized then re-serialized and hash-checked will *fail*.

Switched CapGrantEntry to use BTreeSet instead of HashSet.  Breaking API change!

This isn't a full fix -- there still appears to be problems with multiple Zome init() functions committing things.  But at least CapGrant seems to work more reliably now...

**Comment by pjkundert** [comment]
This test still fails.  If any Zome init function does a create_entry, and there are multiple init functions, the source-chain will be corrupted.  Commenting out one line (the create_entry in one init function) causes the wasm_test tests to succeed.

The "partial" fix for something related to this is to replace HashSet with BTreeSet, so at least the CapGrantEntry is deterministically re-serializable and has a consistent hash.  But, this is not a complete fix...

**Comment by pjkundert** [comment]
Hi, @thedavidmeister; this test still fails, and shows at least 1 serious issue exists in the handling of the source-chain by init functions.  

This will fail the wasm_test (run as indicated above), but commenting out one create_entry line in the init() function causes the tests to pass.

The issue with using HashSet in committed entries seems to be related, and definitely is causing problems -- but isn't the only root cause of this issue; maybe it is completely separate...  Anyway, if you could take a look, this is definitely causing corruption of the source-chain and people will run into this if their init functions do anything.  Thanks!

**Comment by pjkundert** [comment]
@zippy @thedavidmeister I think I have made a breakthrough; after considerable mucking around with removing functionality from my Zomes until they worked -- I have confirmed that *any* Zome function that does more than one `create_entry` will cause an "Unreachable" failure on returning from the API call!  Should be easy for you to test and track down now...

It applies to the `init` call, for example; if it does a CapGrant (which commits and entry), and then does any further create_entry calls, the `init` will fail.  Similarly, if you just put two `create_entry` calls in any Zome function and try to call it, it will fail as it tries to return.

It doesn't appear to trigger the `validate_...` callbacks for the second Entry created, but it does appear to call them for the first one.  So perhaps that is a hint of where an issue might be?

**Comment by pjkundert** [comment]
Another data-point: one of my Zome functions was failing (after it returned!) with an Unreachable error.  I removed the validate_create/update/delete functions, and it started working.  No other changes.

**Comment by pjkundert** [comment]
@zippy OK, this defect seems to be the most urgent for me, now -- and it seems to prevent anyone from writing a non-trivial Zome (one with more than a single `create_entry` call).

Also, the strangeness introduced by the fact that a HashSet or HashMap cannot be re-serialized deterministically is weird.  I would have thought that reconstituting a HashSet/Map from its SerializedBytes should retain the random seed for hashing.  But, it seems to be causing issues...

**Comment by pjkundert** [comment]
@thedavidmeister After merging the latest `develop` with your hdk3 API improvements, we're getting a bit better error context:

```
FATAL PANIC PanicInfo {
    payload: Any,
    message: Some(
        assertion failed: `Ok(Err(WasmError(Host("Wasm error while working with Ribosome: Host(\"An error with entry defs: create_entry\")"))))` does not match `Ok(Ok(ZomeCallResponse :: Ok(_)))`,
    ),
    location: Location {
        file: "crates/holochain/src/core/ribosome/host_fn/call.rs",
        line: 231,
        col: 9,
    },
}
```

This occurs if *any* of the 3 additional error-triggering `create_entry(&Setup...` calls are left uncommented in holochain/crates/test_utils/wasm/wasm_workspace/create_entry/src/lib.rs, and the tests recommended in the initial comment are run.

**Comment by freesig** [comment]
Hi @pjkundert Thanks for creating this PR, I'm having a look into this now.
I'm not surprised about the HashMap/Set not serializing / de-serializing deterministically. Those data structures have no order. Good find.

File: crates/test_utils/wasm/wasm_workspace/create_entry/src/lib.rs [hunk 1]
```diff
@@ -1,5 +1,12 @@
 use hdk3::prelude::*;
 
+#[hdk_entry(
+    id = "setup",
+    required_validations = 5,
+    required_validation_type = "element"
+)]
+struct Setup(String);
```
Comment (freesig) [inline crates/test_utils/wasm/wasm_workspace/create_entry/src/lib.rs:8]: You haven't added this to the entry_defs on line 32.
Needs to be:
```rust
entry_defs![Post::entry_def(), Msg::entry_def(), PrivMsg::entry_def(), Setup::entry_def()];
```

**Review by freesig — CHANGES_REQUESTED** [review]
After fixing the missing entry def on my end everything works.
The BTreeMap is a good fix though. We can either create a new PR with it or if you want you could remove the other unrelated stuff from this PR and we could merge it. Up to you :)

**Comment by pjkundert** [comment]
OK, I've fixed the WASM create_entry test to pass; it still contains the tests to ensure that creating multiple entries in the same Zome API call works.  This is a useful test, I think.  I wonder if there is some kind of logging we can add, that would explain why trying to commit an entry that hasn't been properly set up with a `entry_defs!` macro fails?  Regardless, I think this pull could be merged, as it does contain some good additional tests.

**Comment by pjkundert** [comment]
@freesig Ah, more data...  The problem appears to be caused by or related to the wrong validate_..._entry functions being run, when an entry is created.  My Zome has 5 entries it can create (all properly registered in `entry_defs![`).  However, when I commit a certain entry, sometimes a *different* entry's validate function are executed!  Commenting out the *correct* entry's validate_... functions causes the tests to pass; putting the *correct* entry's validate_... functions back causes the *wrong* entry's validate_... functions to be called.

Further refinement got me some more error context; there appears to be a quite a number of panic! opportunities in the validate workflow, in app_validation_consumer.rs, its hitting this .expect:
```
            if let WorkComplete::Incomplete = app_validation_workflow(
                workspace,
                env.clone().into(),
                &mut trigger_integration,
                conductor_api.clone(),
                network.clone(),
            )
            .await
            .expect("Error running Workflow")
            {
                trigger_self.trigger()
            };

```

and we get this wall of error context:

```
06:59:26 ^[[32minfo^[[39m:
☉☉☉ {{{CONDUCTOR c0}}}
☉ FATAL PANIC PanicInfo {
☉     payload: Any,
☉     message: None,
☉     location: Location {
☉         file: "crates/holochain/src/core/ribosome/host_fn/unreachable.rs",
☉         line: 11,
☉         col: 5,
☉     },
☉ }
☉ thread 'holochain-tokio-thread' panicked at 'internal error: entered unreachable code', crates/holochain/src/core/ribosome/host_fn/unreachable.rs:11:5
☉ stack backtrace:
☉
06:59:26 ^[[32minfo^[[39m:
★★★ {{{CONDUCTOR c0}}}
★    0:     0x55899fbd1d60 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2e8cb764b7fe02e7
★    1:     0x55899fbf9bfc - core::fmt::write::h7a1184eaee6a8644
★
06:59:26 ^[[32minfo^[[39m:
☯☯☯ {{{CONDUCTOR c0}}}
☯    2:     0x55899fbc9632 - std::io::Write::write_fmt::haeeb374d93a67eac
☯    3:     0x55899fbd48bd - std::panicking::default_hook::{{closure}}::hde0cb80358a6920a
☯    4:     0x55899fbd4568 - std::panicking::default_hook::h9b1a691049a0ec8f
☯    5:     0x55899f2e335a - <holochain_lmdb::env::ENVIRONMENTS as core::ops::deref::Deref>::deref::__static_ref_initialize::{{closure}}::h19d6442ddd1e9bb2
☯    6:     0x55899fbd4fc6 - std::panicking::rust_panic_with_hook::h2bdec87b60580584
☯    7:     0x55899eae6594 - std::panicking::begin_panic::{{closure}}::h4b801d609cfb47e7
☯    8:     0x55899eae64ec - std::sys_common::backtrace::__rust_end_short_backtrace::ha7bfad57cadda14e
☯    9:     0x55899eb87eac - std::panicking::begin_panic::hb1b510f490953b2d
☯   10:
06:59:26 ^[[32minfo^[[39m:
☸☸☸ {{{CONDUCTOR c0}}}
☸  0x55899ea6b31c - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h3db25bfaeb021d73
☸   11:     0x55899eb88e05 - std::panicking::try::h70a77346ee917953
☸   12:     0x55899ec71fcb - <FN as wasmer_runtime_core::typed_func::HostFunction<wasmer_runtime_core::typed_func::ExplicitVmCtx,A,Rets>>::to_raw::wrap::h3900a6518a2cb5bd
☸   13:     0x7f3658efaca0 - <unknown>
☸
06:59:26 ^[[32minfo^[[39m:
☮☮☮ {{{CONDUCTOR c0}}}
☮ FATAL PANIC PanicInfo {
☮     payload: Any,
☮     message: Some(
☮         Error running Workflow: AppValidationError(RibosomeError(WasmError(CallError("RuntimeError: \"internal error: entered unreachable code\"")))),
☮     ),
☮     location: Location {
☮         file: "/tmp/nix-build-holochain.drv-0/source/crates/holochain/src/core/queue_consumer/app_validation_consumer.rs",
☮         line: 44,
☮         col: 14,
☮     },
☮ }
☮ thread 'holochain-tokio-thread' panicked at 'Error running Workflow: AppValidationError(RibosomeError(WasmError(CallError("RuntimeError: \"internal error: entered unreachable code\""))))', /tmp/nix-bui$
☮ stack backtrace:
☮    0:     0x55899fbd1d60 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2e8cb764b7fe02e7
☮    1:     0x55899fbf9bfc - core::fmt::write::h7a1184eaee6a8644
☮    2:     0x55899fbc9632 - std::io::Write::write_fmt::haeeb374d93a67eac
☮    3:     0x55899fbd48bd - std::panicking::default_hook::{{closure}
06:59:26 ^[[32minfo^[[39m:
☉☉☉ {{{CONDUCTOR c0}}}
☉ }::hde0cb80358a6920a
☉    4:     0x55899fbd4568 - std::panicking::default_hook::h9b1a691049a0ec8f
☉    5:     0x55899f2e335a - <holochain_lmdb::env::ENVIRONMENTS as core::ops::deref::Deref>::deref::__static_ref_initialize::{{closure}}::h19d6442ddd1e9bb2
☉    6:     0x55899fbd4fc6 - std::panicking::rust_panic_with_hook::h2bdec87b60580584
☉    7:     0x55899fbd4b49 - std::panicking::begin_panic_handler::{{closure}}::h101ca09d9df5db47
☉    8:     0x55899fbd21cc - std::sys_common::backtrace::__rust_end_short_backtrace::h3bb85654c20113ca
☉    9:     0x55899fbd4b09 - rust_begin_unwind
☉   10:     0x55899fbf77f1 - core::panicking
06:59:26 ^[[32minfo^[[39m:
★★★ {{{CONDUCTOR c0}}}
★ ::panic_fmt::h48c31e1e3d550146
★   11:     0x55899fbf75f3 - core::option::expect_none_failed::h6154dc750ae47ade
★   12:     0x55899e73a9c5 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1f16cbe81fbaf32c
★   13:     0x55899e6c0432 - <std::panic::AssertUnwindSafe<F> as
06:59:26 ^[[32minfo^[[39m:
☯☯☯ {{{CONDUCTOR c0}}}
☯  core::ops::function::FnOnce<()>>::call_once::h2fe04f6d58ec3788
☯   14:     0x55899e8b2342 - tokio::runtime::task::harness::Harness<T,S>::poll::he36c30118b1758da
☯   15:     0x55899f9b88a4 - std::thread::local::LocalKey<T>::with::h1855a03b6bcb8e2a
☯   16:     0x55899f9c3832 - tokio::runtime::thread_pool::worker::Context::run_task::h5ea7c41244249c78
06:59:26 ^[[32minfo^[[39m:
☸☸☸ {{{CONDUCTOR c0}}}
☸
☸   17:     0x55899f9c326f - tokio::runtime::thread_pool::worker::Context::run::h7d405af37e7426eb
☸   18:     0x55899f9a4b33 - tokio::macros::scoped_tls::ScopedKey<T>::set::h8952602b5327f8ca
☸   19:     0x55899f9c2a66 - tokio::runtime::thread_pool::worker::run::h8cf96e02fa23a96d
☸   20:
06:59:26 ^[[32minfo^[[39m:
☮☮☮ {{{CONDUCTOR c0}}}
☮    0x55899f9b9b60 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h9384973632b9f087
☮   21:     0x55899f9ac0ef - <std::
06:59:26 ^[[32minfo^[[39m:
☉☉☉ {{{CONDUCTOR c0}}}
☉ panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h7ecdc7e6ae82d3d6
☉   22:     0x55899f9a9c83 - tokio::runtime::task::harness::Harness<T,S>::poll::h47fefc8b938b8183
☉   23:     0x55899f9b7b61 - tokio::runtime::blocking::pool::Inner::run::he051eecddaf6e37f
☉   24:     0x55899f9a5dd0 - tokio::runtime
06:59:26 ^[[32minfo^[[39m:
★★★ {{{CONDUCTOR c0}}}
★ ::context::enter::h07fe93326617b920
★   25:     0x55899f9b8576 - std::sys_common::backtrace::__rust_begin_short_backtrace::hb07420599501ff71
★   26:     0x55899f9a39cb - core::ops::function::FnOnce::call_once{{vtable.shim}}::h919cee9a2592b29e
★   27:     0x55899fbdd00a - std::sys::unix::thread::Thread::new::thread_start::h4afaeade0da13617
★
06:59:26 ^[[32minfo^[[39m:
☯☯☯ {{{CONDUCTOR c0}}}
☯   28:     0x7f36606b5ead - start_thread
☯
06:59:26 ^[[32minfo^[[39m:
☸☸☸ {{{CONDUCTOR c0}}}
☸   29:     0x7f3660913d2f - __clone
☸   30:                0x0 - <unknown>
☸
06:59:26 ^[[32minfo^[[39m:
☮☮☮ [[[CONDUCTOR c0]]]
☮ ^[[2mFeb 21 06:59:26.846^[[0m ^[[34mDEBUG^[[0m ^[[1mwasm_trace^[[0m: holo_profile:zomes/profile/src/lib.rs:191 Initializing holo-profile: Agent: AgentInfo { agent_initial_pubkey: AgentPubKey(uhCAkleVL_$
☮
06:59:26 ^[[32minfo^[[39m:
☉☉☉ [[[CONDUCTOR c0]]]
☉ ^[[2mFeb 21 06:59:26.863^[[0m ^[[34mDEBUG^[[0m ^[[1mwasm_trace^[[0m: holo_profile:zomes/profile/src/lib.rs:264 Getting Profile for GetProfileInput { agent_id: None }
☉
06:59:28 ^[[32minfo^[[39m:
★★★ {{{CONDUCTOR c0}}}
★ FATAL PANIC PanicInfo {
★     payload: Any,
★     message: None,
★     location: Location {
★         file: "crates/holochain/src/core/ribosome/host_fn/unreachable.rs",
★         line: 11,
★         col: 5,
★     },
★ }
★ thread 'holochain-tokio-thread' panicked at 'internal error: entered unreachable code', crates/holochain/src/core/ribosome/host_fn/unreachable.rs:11:5
★ stack backtrace:
★    0:     0x55899fbd1d60 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2e8cb764b7fe02e7
★    1:     0x55899fbf9bfc - core
06:59:28 ^[[32minfo^[[39m:
☯☯☯ {{{CONDUCTOR c0}}}
☯ ::fmt::write::h7a1184eaee6a8644
☯    2:     0x55899fbc9632 - std::io::Write::write_fmt::haeeb374d93a67eac
☯    3:     0x55899fbd48bd - std::panicking::default_hook::{{closure}}::hde0cb80358a6920a
☯    4:     0x55899fbd4568 - std::panicking::default_hook::h9b1a691049a0ec8f
☯    5:     0x55899f2e335a - <holochain_lmdb::env::ENVIRONMENTS as core::ops::deref::Deref>::deref::__static_ref_initialize::{{closure}}::h19d6442ddd1e9bb2
☯    6:     0x55899fbd4fc6 - std::panicking::rust_panic_with_hook::h2bdec87b60580584
☯    7:     0x55899eae6594 - std::panicking::begin_panic::{{closure}}::h4b801d609cfb47e7
☯    8:     0x55899eae64ec - std::sys_common::backtrace::__rust_end_short_backtrace::ha7bfad57cadda14e
☯    9:     0x55899eb87eac - std::panicking::begin_panic::hb1b510f490953b2d
☯   10:     0x55899ea6b31c - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h3db25bfaeb021d73
☯   11:     0x55899eb88e05 - std::panicking::try::h70a77346ee917953
☯   12:     0x55899ec71fcb - <FN
06:59:28 ^[[32minfo^[[39m:
☸☸☸ {{{CONDUCTOR c0}}}
☸ as wasmer_runtime_core::typed_func::HostFunction<wasmer_runtime_core::typed_func::ExplicitVmCtx,A,Rets>>::to_raw::wrap::h3900a6518a2cb5bd
☸   13:     0x7f3658efaca0 - <unknown>
☸
☸
06:59:28 ^[[32minfo^[[39m:
☮☮☮ {{{CONDUCTOR c0}}}
☮ FATAL PANIC PanicInfo {
☮     payload: Any,
☮     message: Some(
☮         Error running Workflow: AppValidationError(RibosomeError(WasmError(CallError("RuntimeError: \"internal error: entered unreachable code\"")))),
☮     ),
☮     location: Location {
☮         file: "/tmp/nix-build-holochain.drv-0/source/crates/holochain/src/core/queue_consumer/app_validation_consumer.rs",
☮         line: 44,
☮         col: 14,
☮     },
☮ }
☮ thread 'holochain-tokio-thread' panicked at 'Error running Workflow: AppValidationError(RibosomeError(WasmError(CallError("RuntimeError: \"internal error: entered unreachable code\""))))', /tmp/nix-bui$
☮ stack backtrace:
☮
06:59:28 ^[[32minfo^[[39m:
☉☉☉ {{{CONDUCTOR c0}}}
☉  0:     0x55899fbd1d60 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2e8cb764b7fe02e7
☉    1:     0x55899fbf9bfc - core::fmt::write::h7a1184eaee6a8644
☉    2:     0x55899fbc9632 - std::io::Write::write_fmt::haeeb374d93a67eac
☉    3:     0x55899fbd48bd - std::panicking::default_hook::{{closure}}::hde0cb80358a6920a
☉    4:     0x55899fbd4568 - std::panicking::default_hook::h9b1a691049a0ec8f
☉    5:     0x55899f2e335a - <holochain_lmdb::env::ENVIRONMENTS as core::ops::deref::Deref>::deref::__static_ref_initialize::{{closure}}
06:59:28 ^[[32minfo^[[39m:
★★★ {{{CONDUCTOR c0}}}
★ ::h19d6442ddd1e9bb2
★    6:     0x55899fbd4fc6 - std::panicking::rust_panic_with_hook::h2bdec87b60580584
★    7:     0x55899fbd4b49 - std::panicking::begin_panic_handler::{{closure}}::h101ca09d9df5db47
★    8:     0x55899fbd21cc - std::sys_common::backtrace::__rust_end_short_backtrace::h3bb85654c20113ca
★    9:     0x55899fbd4b09 - rust_begin_unwind
★   10:     0x55899fbf77f1 - core::panicking::panic_fmt::h48c31e1e3d550146
★   11:     0x55899fbf75f3 - core::option::expect_none_failed::h6154dc750ae47ade
★   12:     0x55899e73a9c5 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h1f16cbe81fbaf32c
★
06:59:28 ^[[32minfo^[[39m:
☯☯☯ {{{CONDUCTOR c0}}}
☯   13:     0x55899e6c0432 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h2fe04f6d58ec3788
☯   14:     0x55899e8b2342 - tokio::runtime::task::harness::Harness<T,S>::poll::he36c30118b1758da
☯   15:     0x55899f9b89b1 - std::thread::local::LocalKey<T>::with::h1855a03b6bcb8e2a
☯   16:     0x55899f9c3832 - tokio::runtime::thread_pool::worker::Context::run_task::h5ea7c41244249c78
☯   17:     0x55899f9c2fbe - tokio::runtime::thread_pool::worker::Context::run::h7d405af37e7426eb
☯   18:     0x55899f9a4b33 - tokio::macros::scoped_tls::ScopedKey<T>::set::
06:59:28 ^[[32minfo^[[39m:
☸☸☸ {{{CONDUCTOR c0}}}
☸ h8952602b5327f8ca
☸   19:     0x55899f9c2a66 - tokio::runtime::thread_pool::worker::run::h8cf96e02fa23a96d
☸   20:     0x55899ebe117f - tokio::runtime::task::core::Core<T,S>::poll::he8a3ffcc972379fc
☸   21:     0x55899ea6ebd2 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::ha10c48f2bf23b3e4
☸   22:     0x55899eb13846 - tokio::runtime::task::harness::Harness<T,S>::poll::h46386970b2004c1b
☸   23:
06:59:28 ^[[32minfo^[[39m:
☮☮☮ {{{CONDUCTOR c0}}}
☮   0x55899f9b7b61 - tokio::runtime::blocking::pool::Inner::run::he051eecddaf6e37f
☮   24:     0x55899f9a5dd0 - tokio::runtime::context::enter::h07fe93326617b920
☮   25:     0x55899f9b8576 - std::sys_common::backtrace::__rust_begin_short_backtrace::hb07420599501ff71
☮   26:     0x55899f9a39cb - core::ops::function::FnOnce::call_once{{vtable.shim}}::h919cee9a2592b29e
☮   27:     0x55899fbdd00a - std::sys::unix::thread::Thread::new::thread_start::h4afaeade0da13617
☮   28:     0x7f36606b5ead - start_thread
☮   29:
06:59:28 ^[[32minfo^[[39m:
☉☉☉ {{{CONDUCTOR c0}}}
☉  0x7f3660913d2f - __clone
☉   30:                0x0 - <unknown>
☉
```

**Comment by pjkundert** [comment]
Commenting out all my validate_create_... functions except one seems to solve the problem, but adding multiple validate_create_... methods to the wasm test cases doesn't seem to cause it to fail.  Still investigating.

**Comment by pjkundert** [comment]
Basically, the only way to get my tests passing is to comment out all validate_... functions, and enable exactly *one* validate_create_... function in one of the Zomes; this sometimes works.  However, it sometimes doesn't -- as if the order of the Zomes in the DNA influences which validate_... functions can be properly made available.  If multiple Zomes have validate_... functions, or if the "wrong" Zome has the validate_... function, then I get an `unreachable` panic.

**Comment by pjkundert** [comment]
This pull is subsumed by [#685], but could be merged and then I would rebase 685 onto develop; your option.

**Comment by thedavidmeister** [comment]
> Basically, the only way to get my tests passing is to comment out all validate_... functions, and enable exactly _one_ validate_create_... function in one of the Zomes; this sometimes works. However, it sometimes doesn't -- as if the order of the Zomes in the DNA influences which validate_... functions can be properly made available. If multiple Zomes have validate_... functions, or if the "wrong" Zome has the validate_... function, then I get an `unreachable` panic.

there are multiple validate functions in deepkey so i'll expect to run into this soonish :sweat_smile:

**Comment by freesig** [comment]
Sorry for falling behind on this PR. What's the current state of it. @thedavidmeister Did you run into this?

**Comment by thedavidmeister** [comment]
@freesig i never saw it before no, @pjkundert 

so i checked this out, and granted it is now pretty stale with merge conflicts but i got this other error:

```
FATAL PANIC PanicInfo {
    payload: Any,
    message: Some(
        elapsed=18446744073709551615; when=223,
    ),
    location: Location {
        file: "/home/thedavidmeister/holochain/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.25/src/time/wheel/mod.rs",
        line: 223,
        col: 9,
    },
}
thread 'tokio-runtime-worker' panicked at 'elapsed=18446744073709551615; when=223', /home/thedavidmeister/holochain/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.25/src/time/wheel/mod.rs:223:9
```

buuuut looking at the error from @pjkundert :

```
★ FATAL PANIC PanicInfo {
★     payload: Any,
★     message: None,
★     location: Location {
★         file: "crates/holochain/src/core/ribosome/host_fn/unreachable.rs",
★         line: 11,
★         col: 5,
★     },
★ }
```

you can see it isn't just any old unreachable code, it is THE `host_fn/unreachable.rs` host function (which no longer exists btw) which indicates that this is a permissions issue

that is to say, you're calling a host function that is routing to `unreachable` on the host because it is being called in a guest callback that isn't allowed to call that host function

given that the permissions were refactored recently on develop i think we have to dust this off and get it up to date with latest develop to see if it is still a problem

**Pull request closed without merging**